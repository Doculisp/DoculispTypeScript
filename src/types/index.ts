// Export all types for external consumption
export * from './types.ast';
export * from './types.astDoculisp';
export * from './types.astProject';
export * from './types.containers';
export * from './types.controller';
export * from './types.document';
export * from './types.docWriter';
export * from './types.fileHandler';
export * from './types.filePath';
export * from './types.general';
export * from './types.includeBuilder';
export * from './types.internal';
export * from './types.sringBuilder';
export * from './types.stringWriter';
export * from './types.structure';
export * from './types.textHelpers';
export * from './types.tokens';
export * from './types.trimArray';
export * from './types.variableTable';
export * from './types.version';