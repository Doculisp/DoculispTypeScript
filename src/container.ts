import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};