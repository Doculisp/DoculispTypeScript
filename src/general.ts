import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};