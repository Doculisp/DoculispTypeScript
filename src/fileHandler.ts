import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};