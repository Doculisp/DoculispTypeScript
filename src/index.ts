#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
