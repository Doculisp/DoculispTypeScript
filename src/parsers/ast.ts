import { AstPart, IAst, IAstParser, ILoad, ITitle } from "../types.ast";
import { IRegisterable } from "../types.containers";
import { ILocation, IUtil, Result } from "../types.general";
import { HandleValue, IInternals, StepParseResult } from "../types.internal";
import { Token, TokenizedDocument } from "../types.tokens";

function headerize(depth: number, value: string): string {
    const id = ''.padStart(depth, '#');
    return `${id} ${value} ${id}`;
}

function trimArray<T>(length: number, values: T[]): T[] {
    for (let index = 0; index < length; index++) {
        values.shift();
    }

    return values;
}

function isSectionMeta(internals: IInternals, util: IUtil): HandleValue<Token[], AstPart> {
    return function(toParse: Token[], starting: ILocation):  StepParseResult<Token[], AstPart> {
        let depth = 0;
        let start: ILocation | undefined;
        let linkText: string | false = false;
        let subTitleText: string | false = false;
        let title: ITitle | false = false;

        function tryParseSectionMeta(input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
            if(input.length < 2) {
                return internals.noResultFound();
            }

            const open = input[0] as Token;
            const atom = input[1] as Token;

            if(open.type !== 'token - open parenthesis') {
                return internals.noResultFound();
            }

            if(atom.type !== 'token - atom' || atom.text !== 'section-meta') {
                return internals.noResultFound();
            }

            //Possible error: (section-meta (section-meta (title something)))
            depth++;
            start = open.location;

            return util.ok({
                type: 'discard',
                location: current.increaseChar(),
                rest: trimArray(2, input),
            });
        }

        function tryParseTitle(input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
            if(!start) {
                return internals.noResultFound();
            }

            if(input.length < 3) {
                return internals.noResultFound();
            }

            const open = input[0] as Token;
            const atom = input[1] as Token;
            const param = input[2] as Token;

            if(open.type !== 'token - open parenthesis') {
                return internals.noResultFound();
            }

            if(atom.type !== 'token - atom' || atom.text !== 'title') {
                return internals.noResultFound();
            }

            if(param.type !== 'token - parameter') {
                return util.fail(`Title at ${open.location.toString()} does not contain title text.`, open.location.documentPath);
            }

            const close = input[3] as Token;
            if(close.type !== 'token - close parenthesis') {
                return internals.noResultFound();
            }

            const link = (
                linkText ? 
                linkText :
                '#' + param.text.toLocaleLowerCase().replaceAll(' ', '_'));

            const label = headerize(open.location.documentDepth, param.text);
            const subtitle: string | undefined = 
                subTitleText ?
                subTitleText :
                undefined;

            title = {
                type: 'ast-title',
                title: param.text,
                subtitle,
                label,
                link,
                documentOrder: start,
            };

            return util.ok({
                type: 'discard',
                rest: trimArray(4, input),
                location: current.increaseChar(),
            });
        }

        function tryParseLink(input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
            if(!start) {
                return internals.noResultFound();
            }

            if(input.length < 3) {
                return internals.noResultFound();
            }

            const open = input[0] as Token;
            const link = input[1] as Token;
            const param = input[2] as Token;

            if(open.type !== 'token - open parenthesis') {
                return internals.noResultFound();
            }

            if(link.type !== 'token - atom' || link.text !== 'link') {
                return internals.noResultFound()
            }

            if(param.type !== 'token - parameter') {
                return util.fail(`the Link Command at ${open.location.toString()} does not have a link text.`, open.location.documentPath);
            }

            const close = input[3] as Token;
            if(close.type !== 'token - close parenthesis') {
                return internals.noResultFound();
            }

            linkText = '#' + param.text;

            if(title) {
                const n: ITitle = {
                    type: 'ast-title',
                    label: title.label,
                    title: title.title,
                    link: linkText,
                    documentOrder: title.documentOrder,
                }

                title = n;
            }

            return util.ok({
                type: 'discard',
                rest: trimArray(4, input),
                location: current.increaseChar(),
            });
        }

        function tryParseSubtitle(input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
            if(!start) {
                return internals.noResultFound();
            }

            if(input.length < 4) {
                return internals.noResultFound();
            }

            const open = input[0] as Token;
            const atom = input[1] as Token;
            const param = input[2] as Token;

            if(open.type !== 'token - open parenthesis') {
                return internals.noResultFound();
            }

            if(atom.type !== 'token - atom' || atom.text !== 'subtitle') {
                return internals.noResultFound();
            }

            if(param.type !== 'token - parameter') {
                return util.fail(`The Subtitle command at ${open.location} does not contain subtitle text.`, open.location.documentPath);
            }

            const close = input[3] as Token;
            if(close.type !== 'token - close parenthesis') {
                return internals.noResultFound();
            }

            const text = headerize(starting.documentDepth + 2, param.text);

            if(title) {
                const nt: ITitle = {
                    type: 'ast-title',
                    title: title.title,
                    subtitle: text,
                    label: title.label,
                    link: title.link,
                    documentOrder: title.documentOrder,
                };

                title = nt;
            }

            subTitleText = text;
            return util.ok({
                type: 'discard',
                rest: trimArray(4, input),
                location: open.location,
            });
        }

        function tryParseClose(input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
            if(!start || depth < 1) {
                return internals.noResultFound();
            }

            if(input.length < 1) {
                return internals.noResultFound();
            }

            const close = input[0] as Token;

            if(close.type !== 'token - close parenthesis') {
                return internals.noResultFound();
            }

            depth--;

            return util.ok({
                type: 'discard',
                rest: trimArray(1, input),
                location: current.increaseChar(),
            });
        }

        function trySubParseLoadable(input: Token[], current: ILocation): StepParseResult<Token[], ILoad> {
            if(!start) {
                return internals.noResultFound();
            }

            if(depth < 2) {
                return internals.noResultFound();
            }

            if(input.length < 4) {
                return internals.noResultFound();
            }

            const open = input[0] as Token;
            const atom = input[1] as Token;
            const param = input[2] as Token;

            if(open.type !== 'token - open parenthesis') {
                return internals.noResultFound();
            }

            if(atom.type !== 'token - atom') {
                return internals.noResultFound();
            }

            if(param.type !== 'token - parameter') {
                return util.fail(`Section command named "${atom.text}" at ${open.location} does not have a section title.`, open.location.documentPath);
            }

            const close = input[3] as Token;
            if(close.type !== 'token - close parenthesis') {
                return internals.noResultFound();
            }

            const load: ILoad = {
                type: 'ast-load',
                sectionLabel: atom.text,
                path: param.text,
                documentOrder: open.location,
                document: false,
            }

            return util.ok({
                type: 'parse result',
                subResult: load,
                rest: trimArray(4, input),
                location: current.increaseChar(),
            });
        }

        function tryParseExternal(input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
            if(!start) {
                return internals.noResultFound();
            }

            if(input.length < 3) {
                return internals.noResultFound();
            }

            const open = input[0] as Token;
            const atom = input[1] as Token;

            if(open.type !== 'token - open parenthesis') {
                return internals.noResultFound();
            }

            if(atom.type !== 'token - atom' || atom.text !== 'external') {
                return internals.noResultFound();
            }

            input = trimArray(2, input);
            depth++;

            const parser = internals.createArrayParser(trySubParseLoadable, tryParseClose);
            const parsed = parser.parse(input, current.increaseChar());

            if(!parsed.success) {
                return parsed;
            }

            const [result, leftovers] = parsed.value;

            if(0 === result.length) {
                return util.fail(`External command at ${open.location} does not contain any section information.`, open.location.documentPath);
            }

            return util.ok({
                type: 'parse group result',
                subResult: result.map(r => { return { type: 'keep', keptValue: r }; }),
                rest: leftovers.remaining,
                location: leftovers.location,
            })
        }

        const parser = internals.createArrayParser(tryParseSectionMeta, tryParseTitle, tryParseClose, tryParseLink, tryParseSubtitle, tryParseExternal);
        const parsed = parser.parse(toParse, starting);

        if(!parsed.success) {
            return parsed;
        }

        if(!start) {
            return internals.noResultFound();
        }

        const [result, leftovers] = parsed.value;

        if(!title) {
            return util.fail(`section-meta atom at ${start.toString()} must contain at least a title.`, starting.documentPath);
        }

        result.unshift(title);

        return util.ok({
            type: 'parse group result',
            subResult: result.map(r => { return { type: 'keep', keptValue: r }; }),
            rest: leftovers.remaining,
            location: leftovers.location,
        });
    };
}

function isHeader(internals: IInternals, util: IUtil): HandleValue<Token[], AstPart> {
    return function (input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
        if(input.length < 3) {
            return internals.noResultFound();
        }

        let open = input[0] as Token;
        let atom = input[1] as Token;
        let param = input[2] as Token;

        if(open.type !== 'token - open parenthesis') {
            return internals.noResultFound();
        }

        if(atom.type !== 'token - atom' || !atom.text.startsWith('#')) {
            return internals.noResultFound();
        }

        if(param.type !== 'token - parameter') {
            return util.fail(`Header at ${open.location.toString()} has no header text.`, open.location.documentPath);
        }

        let close = input[3] as Token;
        if(close.type !== 'token - close parenthesis') {
            return util.fail(`Header at ${open.location.toString()} has unexpected character at ${close.location.toString()}`, close.location.documentPath);
        }

        const match: string = (atom.text.match(/#+/) as any)[0];

        if(match.length !== atom.text.length) {
            return util.fail(`Header at ${open.location.toString()} has invalid syntax.`, open.location.documentPath);
        }

        const part: AstPart = {
            type: 'ast-header',
            depthCount: atom.text.length + open.location.documentDepth,
            text: param.text,
            documentOrder: open.location,
        };

        return util.ok({
            type: 'parse result',
            subResult: part,
            location: current.increaseChar(),
            rest: trimArray(4, input),
        });
    }
}

function isText(internals: IInternals, util: IUtil): HandleValue<Token[], AstPart> {
    return function (input: Token[], current: ILocation): StepParseResult<Token[], AstPart> {
        const token: Token = input[0] as Token;
        if(token.type === 'token - text') {
            return util.ok({
                type: 'parse result',
                subResult: {
                    type: 'ast-write',
                    documentOrder: token.location,
                    value: token.text,
                },
                location: current.increaseChar(),
                rest: trimArray(1, input),
            });
        }
        return internals.noResultFound();
    }
}

function buildAstParser(internals: IInternals, util: IUtil): IAstParser {
    return {
        parse(maybeTokens: Result<TokenizedDocument>): Result<IAst> {
            if(maybeTokens.success){
                const document = maybeTokens.value;
                const parser = internals.createArrayParser(isText(internals, util), isHeader(internals, util), isSectionMeta(internals, util));
                const parsed = parser.parse(document.tokens, util.toLocation(document.projectLocation, 1, 1));
                
                if(parsed.success) {
                    const [result, _leftovers] = parsed.value;

                    if(0 < result.length) {
                        return util.ok({
                            projectLocation: document.projectLocation,
                            section: {
                                type: 'ast-section',
                                ast: result,
                                documentOrder: util.toLocation(document.projectLocation, 1, 1),
                            },
                        });
                    }

                    return util.ok({
                        projectLocation: document.projectLocation,
                        section: { type: 'ast-empty' },
                    });
                }

                return parsed;
            }

            return maybeTokens;
        }
    };
}

const astParser: IRegisterable = {
    builder: (internals: IInternals, util: IUtil) => buildAstParser(internals, util),
    name: 'astParse',
    singleton: true,
    dependencies: ['internals', 'util']
};

export {
    astParser,
};