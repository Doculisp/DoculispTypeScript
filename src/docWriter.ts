import { AstBulletStyle, AstPart, IAst, IHeader, ILoad, ISectionWriter, ITableOfContents, ITitle, IWrite } from "./types.ast";
import { IRegisterable } from "./types.containers";
import { IDocWriter } from "./types.docWriter";
import { IUtil, Result } from "./types.general";

function buildWriter(util: IUtil): IDocWriter {
    function getLabel(style: AstBulletStyle, load: ILoad): string {
        switch (style) {
            case 'bulleted':
            case 'numbered':
            case 'unlabeled':
                return '';
            default:
                return load.sectionLabel;
        }
    }

    function getNumberedBullet(): (label: string, text: string) => string {
        let cnt = 1;
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            const result = `${cnt}. ${lbl}${text}`;
            cnt++;
            return result;
        }

        return bullet;
    }

    function getBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `* ${lbl}${text}`;
        }

        return bullet;
    }

    function getNoBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `${lbl}${text}`;
        }
    }

    function getBulletBuilder(style: AstBulletStyle): (label: string, text: string) => string {
        switch (style) {
            case 'bulleted':
            case 'bulleted-labeled':
                return getBullet();
            case 'numbered':
            case 'numbered-labeled':
                return getNumberedBullet();
            default:
                return getNoBullet();
        }
    }

    function compileToc(astToc: ITableOfContents, loads: ILoad): string {
        if(astToc.bulletStyle === 'no-table') {
            return '';
        }
    }

    function compileContentLocation(loads: ILoad[]): string {
        const parts: string[] = [];

        loads.forEach(load => {
            if(load.document) {
                const section = compileSection(load.document);
                parts.push(section);
            }
        });

        return `${parts.join('\n')}`;
    }

    function compileHeader(astHeader: IHeader): string {
        const padding = '#'.repeat(astHeader.depthCount);
        return `${padding} ${astHeader.text} ${padding}`;
    }

    function compileTitle(astTitle: ITitle): string {
        return astTitle.label;
    }

    function compileWrite(astWrite: IWrite): string {
        return astWrite.value;
    }

    function compilePart(astPart: AstPart) : string {
        astPart.
    }

    function compileSection(astSection: ISectionWriter): string {
        const parts: string[] = [];

        
    }

    function compile(astResult: Result<IAst>): Result<string> {
        if(!astResult.success) {
            return astResult;
        }

        const ast = astResult.value;

        if(ast.section.type === 'ast-empty') {
            return util.ok('');
        }

        throw new Error("Not yet Implemented");
    }

    function write(path: string, astResult: Result<IAst>): Result<string> {
        throw new Error('Not yet Implemented');
    }

    return {
        compile,
        write,
    };
}

const documentWriter: IRegisterable = {
    builder: (util: IUtil) => buildWriter(util),
    name: 'docWriter',
    singleton: true,
    dependencies: ['util', 'fileHandler']
};

export {
    documentWriter,
};