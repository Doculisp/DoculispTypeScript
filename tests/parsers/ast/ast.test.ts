import { Options } from "approvals/lib/Core/Options";
import { configure } from "approvals/lib/config";
import { getVerifier } from "../../tools";
import { container } from "../../../src/container";
import { ITestableContainer } from "../../../src/types.containers";
import { IAst, IAstParser } from '../../../src/types.ast'
import { IFail, IProjectLocation, ISuccess, IUtil, Result } from "../../../src/types.general";
import { TokenFunction, TokenizedDocument } from "../../../src/types.tokens";
import { DocumentParser } from "../../../src/types.document";

function buildLocation(path: string, depth: number, index: number) : IProjectLocation {
    return {
        documentPath: path,
        documentDepth: depth,
        documentIndex: index,
    };
}

describe('ast', () => {
    let environment: ITestableContainer = undefined as any;
    let parser: IAstParser = undefined as any;
    let verifyAsJson: (data: any, options?: Options) => void;
    let ok: (successfulValue: any) => ISuccess<any> = undefined as any;
    let fail: (message: string, documentPath: string) => IFail = undefined as any;
    let util: IUtil = undefined as any;
    let toResult: (text: string, projectLocation: IProjectLocation) => Result<IAst> = undefined as any;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    beforeEach(() => {
        environment = container.buildTestable();
        parser = environment.buildAs<IAstParser>('astParse');
        util = environment.buildAs<IUtil>('util');
        let document = environment.buildAs<DocumentParser>('documentParse');
        let tokenizer = environment.buildAs<TokenFunction>('tokenizer');

        toResult = (text: string, projectLocation: IProjectLocation) => {
            const docResult = document(text, projectLocation);
            const tokens = tokenizer(docResult);
            return parser.parse(tokens);
        };
        
        ok = util.ok;
        fail = util.fail;
    });

    describe('basic functionality', () => {
        test('should return failure if given failure', () => {
            const failure = fail('this is a document failure', 'Z:/mybad.dlisp');
    
            const result = parser.parse(failure);
    
            verifyAsJson(result);
        });
    
        test('should return an empty ast if there was no tokens', () => {
            const tokens: Result<TokenizedDocument> = ok({
                projectLocation: buildLocation('A:/empty/doc.md', 4, 10),
                tokens: []
            });
    
            const result = parser.parse(tokens);
    
            verifyAsJson(result);
        });
    
        test('should parse a text token', () => {
            const projectLocation = buildLocation('T:/ext/only.md', 2, 9);
            const tokens: Result<TokenizedDocument> = ok({
                projectLocation: projectLocation,
                tokens: [
                    {
                        type: 'token - text',
                        location: util.toLocation(projectLocation, 2, 1),
                        text: 'Some text',
                    }
                ],
            });
    
            const result = parser.parse(tokens);
    
            verifyAsJson(result);
        });
    
        test('should parse multiple text tokens', () => {
            const projectLocation = buildLocation('T:/ext/only.md', 4, 8);
            const tokens: Result<TokenizedDocument> = ok({
                projectLocation: projectLocation,
                tokens: [
                    {
                        type: 'token - text',
                        location: util.toLocation(projectLocation, 1, 1),
                        text: 'Intro text',
                    },
                    {
                        type: 'token - text',
                        location: util.toLocation(projectLocation, 5, 1),
                        text: 'Text after some comment',
                    }
                ],
            });
    
            const result = parser.parse(tokens);
    
            verifyAsJson(result);
        });
    });

    describe('lisp', () => {
        test('should simple lisp tokens', () => {
            const contents = `<!--
(dl (# My heading))
-->`;
            const result = toResult(contents, buildLocation('S:/ome/file.md', 2, 1));
    
            verifyAsJson(result);
        });
    
        test('should not parse a bad header', () => {
            const contents = `<!--
(dl (#head My heading))
-->`;
            const result = toResult(contents, buildLocation('S:/ome/file.md', 3, 2));
    
            verifyAsJson(result);
        });
    
        test('should not parse a header without a parameter', () => {
            const contents = `<!--
(dl (#))
-->`;
            const result = toResult(contents, buildLocation('S:/ome/file.md', 2, 3));
    
            verifyAsJson(result);
        });

        describe('section-meta', () => {
            describe('title', () => {
                test('should parse a title', () => {
                    const contents = `
(section-meta
    (title My Cool Document)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 7));
            
                    verifyAsJson(result);
                });
            
                test('should not parse a title without a parameter', () => {
                    const contents = `
(section-meta
    (title)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 1));
            
                    verifyAsJson(result);
                });

                test('should parse the link', () => {
                    const contents = `
(section-meta
    (title My cool title✨)
    (link my_cool_title)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 10));
        
                    verifyAsJson(result);
                });

                test('should parse the link if it comes before the title', () => {
                    const contents = `
(section-meta
    (link my_cool_title)
    (title My cool title✨)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 10));
        
                    verifyAsJson(result);
                });

                test('should not parse a link with no parameter', () => {
                    const contents = `
(section-meta
    (link)
    (title My cool title✨)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 10));
                            
                    verifyAsJson(result);
                });

                test('should parse the subtitle command', () => {
                    const contents = `
(section-meta
    (title My cool title)
    (subtitle This is information)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 4, 4));
        
                    verifyAsJson(result);
                });

                test('should parse the subtitle before title command', () => {
                    const contents = `
(section-meta
    (subtitle This is information)
    (title My cool title)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 2, 7));
        
                    verifyAsJson(result);
                });

                test('should not parse a subtitle without a parameter', () => {
                    const contents = `
(section-meta
    (title My cool title)
    (subtitle)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 10));
        
                    verifyAsJson(result);
                });
            });
        });
    });
});