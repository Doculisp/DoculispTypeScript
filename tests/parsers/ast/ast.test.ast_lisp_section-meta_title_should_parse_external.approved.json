{
  "success": true,
  "value": {
    "projectLocation": {
      "documentDepth": 1,
      "documentIndex": 4,
      "documentPath": "main.dlisp"
    },
    "section": {
      "ast": [
        {
          "documentOrder": {
            "char": 1,
            "documentDepth": 1,
            "documentIndex": 4,
            "documentPath": "main.dlisp",
            "line": 1
          },
          "label": "# Doculisp a short description #",
          "link": "#doculisp_a_short_description",
          "title": "Doculisp a short description",
          "type": "ast-title"
        },
        {
          "document": false,
          "documentOrder": {
            "char": 9,
            "documentDepth": 1,
            "documentIndex": 4,
            "documentPath": "main.dlisp",
            "line": 4
          },
          "path": "./doculisp.md",
          "sectionLabel": "section",
          "type": "ast-load"
        },
        {
          "document": false,
          "documentOrder": {
            "char": 9,
            "documentDepth": 1,
            "documentIndex": 4,
            "documentPath": "main.dlisp",
            "line": 5
          },
          "path": "./section-meta.md",
          "sectionLabel": "section",
          "type": "ast-load"
        }
      ],
      "documentOrder": {
        "char": 1,
        "documentDepth": 1,
        "documentIndex": 4,
        "documentPath": "main.dlisp",
        "line": 1
      },
      "type": "ast-section"
    }
  }
}
